core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,238 @@
+   * once the size reaches hardlimit, blocking starts to happen.
+   */
+  private final int mHardLimit;
+  private final int mSoftLimit;
+  private final Function<? super K, ? extends ReentrantReadWriteLock> mDefaultLoader;
+  private Iterator<Map.Entry<K, ValNode>> mIterator;
+  private final Lock mEvictLock;
+
+  private long mLastLogTime;
+
+  /**
+   * Constructor for a lock cache.
+   *
+   * @param defaultLoader specify a function to generate a value based on a key
+   * @param initialSize initial size of the cache
+   * @param maxSize maximum size of the cache
+   * @param concurrencyLevel concurrency level of the cache
+   */
+  public LockCache(@Nullable Function<? super K, ? extends ReentrantReadWriteLock> defaultLoader, int initialSize,
+      int maxSize, int concurrencyLevel) {
+    mDefaultLoader = defaultLoader;
+
+    mHardLimit = maxSize;
+    mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
+    mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+    mLastLogTime = System.currentTimeMillis();
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread won the race as the evictor
+        while (numToEvict > 0) {
+
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
+          ValNode candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * get the value from the cache.
+   *
+   * @param key the key to look up the cache
+   * @return the value contained in the cache, if it is already in cache,
+   * otherwise generate an entry based on the loader
+   */
+
+  public LockResource get(final K key, LockResource.LockMode mode) {
+    ValNode valNode = getValNode(key, true);
+    ReentrantReadWriteLock lock = valNode.mValue;
+    switch (mode) {
+      case READ:
+        return new RefCountLockResource(lock.readLock(), valNode.mRefCount);
+      case WRITE:
+        return new RefCountLockResource(lock.writeLock(), valNode.mRefCount);
+      default:
+        throw new IllegalStateException("Unknown lock mode: " + mode);
+    }
+  }
+
+  public ReentrantReadWriteLock getRawLock(final K key) {
+    ValNode valNode = getValNode(key, false);
```

Comment:
I think we can just return `mCache.getOrDefault(key, new ValNode(new ReentrantReadWriteLock())).mValue` without needing to add the `refCount` parameter to `getValNode`

Since this method is for testing purposes only, I'd like to avoid adding complexity to `getValNode` just for the sake of this method.

Reply:
done